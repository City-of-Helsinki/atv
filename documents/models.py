from django.conf import settings
from django.contrib.postgres.indexes import GinIndex
from django.core.serializers.json import DjangoJSONEncoder
from django.db import models
from django.urls import reverse
from django.utils.translation import gettext_lazy as _
from encrypted_fields import fields

from atv.exceptions import MaximumFileSizeExceededException
from documents.utils import get_attachment_file_path
from documents.validators import BusinessIDValidator
from services.models import Service
from utils.models import TimestampedModel, UUIDModel


class Attachment(TimestampedModel):
    document = models.ForeignKey(
        "Document",
        on_delete=models.CASCADE,
        verbose_name=_("attachment"),
        related_name="attachments",
        help_text=_("Document to which the file is attached."),
    )
    media_type = models.CharField(
        max_length=255,
        default="",
        verbose_name=_("media type"),
        help_text=_("The media type of the attachment."),
    )
    size = models.PositiveIntegerField(
        verbose_name=_("size"),
        help_text=_("Size of the attachment in bytes."),
    )
    filename = models.CharField(
        max_length=255,
        verbose_name=_("filename"),
        help_text=_("The original filename of the attachment."),
    )
    # TODO: The file is not being encrypted yet
    file = models.FileField(
        upload_to=get_attachment_file_path,
        verbose_name=_("file"),
        help_text=_("Encrypted file."),
    )

    class Meta:
        verbose_name = _("attachment")
        verbose_name_plural = _("attachments")
        default_related_name = "attachments"

    def __str__(self):
        return f"Attachment {self.pk}"

    @property
    def uri(self):
        return reverse("documents-attachments-detail", args=[self.document.id, self.id])

    def clean(self):
        if self.file.size > settings.MAX_FILE_SIZE:
            raise MaximumFileSizeExceededException()

    def save(self, *args, **kwargs):
        self.size = self.file.size
        self.filename = self.file.name

        self.full_clean()

        super().save(*args, **kwargs)


class Document(UUIDModel, TimestampedModel):
    service = models.ForeignKey(
        Service,
        on_delete=models.PROTECT,
        verbose_name=_("service"),
        help_text=_("The service which owns this document."),
    )
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        blank=True,
        null=True,
        on_delete=models.PROTECT,
        verbose_name=_("user"),
        help_text=_("The creator/owner of this document."),
    )
    business_id = models.CharField(
        max_length=9,
        blank=True,
        validators=[BusinessIDValidator()],
        verbose_name=_("business ID"),
        help_text=_("The business ID of the organization which owns this document."),
    )
    transaction_id = models.CharField(
        max_length=255,
        blank=True,
        verbose_name=_("transaction ID"),
        help_text=_(
            "Transaction identifier given by the owning service. Could be e.g. a UUID."
        ),
    )
    tos_function_id = models.CharField(
        max_length=32,
        verbose_name=_("TOS function ID"),
        help_text=_(
            "UUID without dashes. Should correspond with a Function instance "
            "(e.g. the id from https://api.hel.fi/helerm/v1/function/eb30af1d9d654ebc98287ca25f231bf6/) "
            "which is applied to the stored document when considering storage time."
        ),
    )
    tos_record_id = models.CharField(
        max_length=32,
        verbose_name=_("TOS record ID"),
        help_text=_(
            "UUID without dashes. Should correspond to a record ID "
            "(e.g. records[].id from https://api.hel.fi/helerm/v1/function/eb30af1d9d654ebc98287ca25f231bf6/) "
            "within a Function instance which is applied to the stored document when "
            "considering storage time."
        ),
    )
    draft = models.BooleanField(
        default=False,
        verbose_name=_("draft"),
        help_text=_(
            "Is this document a draft or not. Drafts can be modified by a user."
        ),
    )
    locked_after = models.DateTimeField(
        blank=True,
        null=True,
        verbose_name=_("locked after"),
        help_text=_(
            "Date and time after which this document cannot be modified, except for "
            "deleting. This field should be filled by the calling service if it knows "
            "e.g. that a certain application has a deadline."
        ),
    )
    metadata = models.JSONField(
        default=dict,
        blank=True,
        # null=True,  # TODO should we allow null?
        encoder=DjangoJSONEncoder,
        verbose_name=_("metadata"),
        help_text=_(
            "Key-value pairs given by the calling service. These fields should enable "
            "the service to store some relevant information which it can use to "
            "filter/sort documents, e.g. the handler of the document."
        ),
    )
    content = fields.EncryptedTextField(
        help_text=_("Encrypted content of the document.")
    )
    status = models.CharField(
        max_length=255,
        blank=True,
        help_text=_(
            "Status information given by the owning service. "
            "Could be e.g. some constant string."
        ),
    )
    type = models.CharField(
        max_length=255,
        blank=True,
        verbose_name=_("type"),
        help_text=_(
            "Type information given by the owning service. "
            "Could be e.g. the type of the document."
        ),
    )

    class Meta:
        verbose_name = _("document")
        verbose_name_plural = _("documents")
        default_related_name = "documents"
        indexes = [
            models.Index(fields=["created_at"], name="document_created_at_idx"),
            models.Index(fields=["updated_at"], name="document_updated_at_idx"),
            models.Index(fields=["business_id"], name="document_business_id_idx"),
            models.Index(fields=["transaction_id"], name="document_transaction_id_idx"),
            models.Index(fields=["draft"], name="document_draft_idx"),
            models.Index(fields=["locked_after"], name="document_locked_after_idx"),
            GinIndex(fields=["metadata"], name="document_metadata_idx"),
            models.Index(fields=["status"], name="document_status_idx"),
            models.Index(fields=["type"], name="document_type_idx"),
        ]

    def __str__(self):
        return f"Document {self.pk}"
