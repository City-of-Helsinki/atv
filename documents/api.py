from rest_framework import status, viewsets
from rest_framework.response import Response


class DocumentViewSet(viewsets.ViewSet):
    def create(self, request):
        return Response({"status": "CREATED"}, status=status.HTTP_201_CREATED)
