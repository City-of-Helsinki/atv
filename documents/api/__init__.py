from .viewsets import AttachmentViewSet, DocumentViewSet

__all__ = [
    "AttachmentViewSet",
    "DocumentViewSet",
]
