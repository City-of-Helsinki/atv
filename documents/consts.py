VALID_OWNER_PATCH_FIELDS = {"draft", "metadata", "content", "attachments"}
