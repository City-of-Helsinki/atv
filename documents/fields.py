import json
from io import BytesIO

from Crypto.Cipher import AES
from django.db import models
from encrypted_fields.fields import EncryptedFieldMixin

from atv.settings import FIELD_ENCRYPTION_KEYS


class EncryptedJSONField(EncryptedFieldMixin, models.JSONField):
    def decrypt(self, value):
        text = super().decrypt(value)
        return json.loads(text)

    def get_db_prep_save(self, value, connection):
        if self.empty_strings_allowed and value == bytes():
            value = ""
        # Instead of using db_prep_value use json.dumps() to get a json formatted object
        value = json.dumps(value)
        if value is not None:
            encrypted_value = self.encrypt(value)
            return connection.Database.Binary(encrypted_value)


class EncryptedFileField(models.FileField):
    def pre_save(self, model_instance, add):
        file = getattr(model_instance, self.attname)
        if file and not file._committed:
            # Commit the file to storage prior to saving the model
            file.save(file.name, self.encrypt_file(file.file), save=False)
        return file

    @staticmethod
    def encrypt_file(file):
        data_to_encrypt = file.read()
        cipher = AES.new(bytes.fromhex(FIELD_ENCRYPTION_KEYS[0]), AES.MODE_GCM)
        nonce = cipher.nonce
        cypher_text, tag = cipher.encrypt_and_digest(data_to_encrypt)
        return BytesIO(nonce + tag + cypher_text)
