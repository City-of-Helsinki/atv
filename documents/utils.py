from io import BytesIO

from Cryptodome.Cipher import AES
from django.conf import settings
from django.core.files import File
from pyclamd import pyclamd

from atv.exceptions import MaliciousFileException
from atv.settings import FIELD_ENCRYPTION_KEYS


def get_attachment_file_path(instance, filename):
    """File will be uploaded to MEDIA_ROOT/ATTACHMENT_MEDIA_DIR/<document_id>/<filename>"""
    return f"{settings.ATTACHMENT_MEDIA_DIR}/{instance.document.id}/{filename}"


def get_document_attachment_directory_path(instance):
    """Get the root directory for a document's attachments.

    :type instance: documents.models.Document
    """
    return f"{settings.ATTACHMENT_MEDIA_DIR}/{instance.id}/"


def get_decrypted_file(file, file_name):
    nonce = file[:16]
    # Perform same nonce checks here as Pycryptodome, so we can raise a more user
    # friendly error message
    if not isinstance(nonce, (bytes, bytearray, memoryview)) or len(nonce) != 16:
        raise ValueError("Data is corrupted.")
    tag = file[16:32]
    cypher_text = file[32:]
    counter = 0
    num_keys = len(FIELD_ENCRYPTION_KEYS)
    while counter < num_keys:
        cipher = AES.new(
            bytes.fromhex(FIELD_ENCRYPTION_KEYS[counter]), AES.MODE_GCM, nonce=nonce
        )
        try:
            plaintext = cipher.decrypt_and_verify(cypher_text, tag)
        except ValueError:
            counter += 1
            continue
        return File(BytesIO(plaintext), name=file_name)
    raise ValueError("AES Key incorrect or data is corrupted")


# TODO: Consider scanning files on download as well to improve chance of catching most recent threats to protect users
#   if clamav virus databases didn't include the virus' profile at the time of upload
# Note this function is mocked in testing because there is no clamav connection during pipeline testing
def virus_scan_attachment_file(file_data):
    cd = pyclamd.ClamdNetworkSocket(host=settings.CLAMAV_HOST)
    if cd.scan_stream(file_data) is not None:
        raise MaliciousFileException
