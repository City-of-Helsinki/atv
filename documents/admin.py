from django.contrib import admin
from django.urls import reverse
from django.utils.html import format_html
from django.utils.translation import gettext_lazy as _

from documents.models import Attachment, Document


@admin.register(Document)
class DocumentAdmin(admin.ModelAdmin):
    list_display = ("pk", "get_service_name", "draft", "created_at", "updated_at")
    search_fields = ("service__name", "type", "status", "business_id")
    list_filter = ("draft",)
    autocomplete_fields = ("service", "user")
    list_select_related = ("service",)

    @admin.display(description=_("service"), ordering="-service__name")
    def get_service_name(self, obj):
        return obj.service.name


@admin.register(Attachment)
class AttachmentAdmin(admin.ModelAdmin):
    list_display = (
        "pk",
        "filename",
        "media_type",
        "get_size_in_mb",
        "get_document",
        "created_at",
        "updated_at",
    )
    search_fields = (
        "filename",
        "document__id",
    )
    list_filter = ("media_type",)
    autocomplete_fields = ("document",)
    readonly_fields = (
        "filename",
        "media_type",
        "size",
    )

    @admin.display(description=_("size (MB)"))
    def get_size_in_mb(self, obj):
        if not obj.size:
            return "-"
        size_mb = round(float(obj.size) / (1024 ** 2), 2)
        return f"{size_mb}MB"

    @admin.display(description=_("Document"))
    def get_document(self, obj):
        link = reverse("admin:documents_document_change", args=[obj.document.id])
        return format_html('<a href="{}">{}</a>', link, str(obj.document))
