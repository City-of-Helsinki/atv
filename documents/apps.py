from django.apps import AppConfig
from django.utils.translation import gettext_lazy as _


class DocumentsConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "documents"
    verbose_name = _("documents")

    def ready(self):
        import documents.signals  # noqa
