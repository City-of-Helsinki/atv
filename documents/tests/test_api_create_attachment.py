from unittest import mock

import pytest
from django.core.files.uploadedfile import SimpleUploadedFile
from django.test import override_settings
from freezegun import freeze_time
from guardian.shortcuts import assign_perm
from rest_framework import status
from rest_framework.reverse import reverse

from atv.tests.factories import GroupFactory
from audit_log.enums import Role
from audit_log.models import AuditLogEntry
from documents.models import Attachment
from documents.tests.factories import DocumentFactory
from services.enums import ServicePermissions
from services.tests.utils import get_user_service_client
from utils.exceptions import get_error_response


@pytest.fixture
def document_data():
    return {
        "file": SimpleUploadedFile(
            "document1.pdf",
            b"file_content",
            content_type="application/pdf",
        )
    }


# OWNER-RELATED ACTIONS


@freeze_time("2021-06-30T12:00:00+03:00")
def test_create_and_retrieve_attachment(user, service, document_data, snapshot):
    api_client = get_user_service_client(user, service)
    document = DocumentFactory(
        id="5209bdd0-e626-4a7d-aa4d-73aaf961a93f",
        user=user,
        service=service,
        draft=True,
    )
    with mock.patch(
        "documents.serializers.attachment.virus_scan_attachment_file", return_value=None
    ):

        response = api_client.post(
            reverse("documents-attachments-list", args=[document.id]), document_data
        )

    assert response.status_code == status.HTTP_201_CREATED

    body = response.json()
    attachment_id = body.pop("id")

    assert document.attachments.count() == 1
    assert document.attachments.first().id == attachment_id

    response = api_client.get(
        reverse("documents-attachments-detail", args=[document.id, attachment_id])
    )
    assert response.status_code == status.HTTP_200_OK

    snapshot.assert_match(body)


@freeze_time("2021-06-30T12:00:00+03:00")
def test_create_attachment_other_document(user, service):
    document = DocumentFactory(service=service)
    api_client = get_user_service_client(user, service)

    response = api_client.post(
        reverse("documents-attachments-list", args=[document.id]), {}
    )

    body = response.json()
    assert response.status_code == status.HTTP_404_NOT_FOUND
    assert body == get_error_response(
        "NOT_FOUND",
        "No Document matches the given query.",
    )


@freeze_time("2021-06-30T12:00:00+03:00")
def test_create_attachment_document_not_draft(user, service):
    document = DocumentFactory(
        id="5209bdd0-e626-4a7d-aa4d-73aaf961a93f",
        user=user,
        service=service,
        draft=False,
    )
    api_client = get_user_service_client(user, service)

    response = api_client.post(
        reverse("documents-attachments-list", args=[document.id]),
        {},
    )

    body = response.json()
    assert response.status_code == status.HTTP_403_FORBIDDEN
    assert body == get_error_response(
        "DOCUMENT_LOCKED",
        "Unable to modify document - it's no longer a draft",
    )


@freeze_time("2021-06-30T12:00:00+03:00")
def test_create_attachment_missing_document_id(user, service, document_data, snapshot):
    api_client = get_user_service_client(user, service)

    response = api_client.post(
        reverse("documents-attachments-list", args=[None]), document_data
    )

    body = response.json()
    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert body == get_error_response(
        "MISSING_PARAMETER",
        "Missing parameter: document_id.",
    )


@override_settings(MAX_FILE_SIZE=50)
def test_create_document_file_limit(user, service, snapshot, settings):
    api_client = get_user_service_client(user, service)
    document = DocumentFactory(
        id="5209bdd0-e626-4a7d-aa4d-73aaf961a93f",
        user=user,
        service=service,
        draft=True,
    )
    data = {
        "file": SimpleUploadedFile(
            "document1.pdf",
            b"x" * 10000,
            content_type="application/pdf",
        )
    }

    response = api_client.post(
        reverse("documents-attachments-list", args=[document.id]), data
    )

    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert Attachment.objects.count() == 0

    body = response.json()
    assert body == get_error_response(
        "MAXIMUM_FILE_SIZE_EXCEEDED",
        "Cannot upload files larger than 0.0 Mb: 0.01 Mb",
    )


# STAFF-RELATED ACTION


@freeze_time("2021-06-30T12:00:00+03:00")
@pytest.mark.parametrize("has_permission", [True, False])
def test_create_attachment_staff(
    user, service, document_data, snapshot, has_permission
):
    api_client = get_user_service_client(user, service)

    group = GroupFactory()
    assign_perm(ServicePermissions.VIEW_DOCUMENTS.value, group, service)
    assign_perm(ServicePermissions.VIEW_ATTACHMENTS.value, group, service)
    if has_permission:
        assign_perm(ServicePermissions.ADD_ATTACHMENTS.value, group, service)
    user.groups.add(group)

    document = DocumentFactory(
        id="5209bdd0-e626-4a7d-aa4d-73aaf961a93f",
        service=service,
        draft=True,
    )
    with mock.patch(
        "documents.serializers.attachment.virus_scan_attachment_file", return_value=None
    ):
        response = api_client.post(
            reverse("documents-attachments-list", args=[document.id]), document_data
        )

    if has_permission:
        assert response.status_code == status.HTTP_201_CREATED
        body = response.json()
        attachment_id = body.pop("id")
        assert document.attachments.count() == 1
        assert document.attachments.first().id == attachment_id
        snapshot.assert_match(body)
    else:
        body = response.json()
        assert document.attachments.count() == 0
        assert response.status_code == status.HTTP_403_FORBIDDEN
        assert body == get_error_response(
            "PERMISSION_DENIED",
            "You do not have permission to perform this action.",
        )


# OTHER STUFF


@pytest.mark.parametrize(
    "ip_address", ["213.255.180.34", "2345:0425:2CA1::0567:5673:23b5"]
)
def test_audit_log_is_created_when_creating(
    user, service, document_data, snapshot, ip_address
):
    api_client = get_user_service_client(user, service)
    document = DocumentFactory(
        id="5209bdd0-e626-4a7d-aa4d-73aaf961a93f",
        user=user,
        service=service,
        draft=True,
    )
    with mock.patch(
        "documents.serializers.attachment.virus_scan_attachment_file", return_value=None
    ):
        response = api_client.post(
            reverse("documents-attachments-list", args=[document.id]),
            document_data,
            HTTP_X_FORWARDED_FOR=ip_address,
        ).json()

    assert document.attachments.count() == 1
    assert (
        AuditLogEntry.objects.filter(
            message__audit_event__target__type="Attachment",
            message__audit_event__target__id=str(response["id"]),
            message__audit_event__operation="CREATE",
            message__audit_event__actor__service=service.name,
            message__audit_event__actor__role=Role.USER,
            message__audit_event__actor__ip_address=ip_address,
        ).count()
        == 1
    )
