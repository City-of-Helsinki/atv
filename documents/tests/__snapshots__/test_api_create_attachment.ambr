# serializer version: 1
# name: test_create_and_retrieve_attachment
  dict({
    'created_at': '2021-06-30T12:00:00+03:00',
    'filename': 'document1.pdf',
    'href': 'http://testserver/v1/documents/5209bdd0-e626-4a7d-aa4d-73aaf961a93f/attachments/1/',
    'media_type': 'application/pdf',
    'size': 12,
    'updated_at': '2021-06-30T12:00:00+03:00',
  })
# ---
# name: test_create_attachment_staff[True]
  dict({
    'created_at': '2021-06-30T12:00:00+03:00',
    'filename': 'document1.pdf',
    'href': 'http://testserver/v1/documents/5209bdd0-e626-4a7d-aa4d-73aaf961a93f/attachments/2/',
    'media_type': 'application/pdf',
    'size': 12,
    'updated_at': '2021-06-30T12:00:00+03:00',
  })
# ---
